### PR TITLE
[BEAM-11075] Go SDK SideInput load tests

### PR DESCRIPTION
**Pull request by tszerszen**
This PR adds SideInput load tests for Go SDK.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by tszerszen** [comment]
R: @kamilwu ?

**Comment by tszerszen** [comment]
Run Spotless PreCommit

File: .test-infra/jenkins/job_LoadTests_SideInput_Flink_Go.groovy [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+def fromTemplate = { mode, name, id, testSpecificOptions ->
+  [
+    title          : "SideInput Go Load test: ${name}",
+    test           : 'sideinput',
+    runner         : CommonTestProperties.Runner.FLINK,
+    pipelineOptions: [
+      job_name           : "load-tests-go-flink-${mode}-sideinput-${id}-${now}",
+      influx_namespace   : 'flink',
+      influx_measurement : "go_${mode}_sideinput_${id}_${now}",
```
Comment (kamilwu) [inline .test-infra/jenkins/job_LoadTests_SideInput_Flink_Go.groovy:35]: Why `${now}`? We would quickly end up with tons of different measurements in the database with only one element.

File: sdks/go/test/load/sideinput/sideinput.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+)
+
+var (
+	accessPercentage = flag.Int(
+		"access_percentage",
+		100,
+		"Specifies the percentage of elements in the side input to be accessed.")
+	syntheticSourceConfig = flag.String(
+		"input_options",
+		"{"+
```
Comment (kamilwu) [inline sdks/go/test/load/sideinput/sideinput.go:36]: Please do not provide any default values for this parameter, as it should be required.

File: sdks/go/test/load/sideinput/sideinput.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+				break
+			}
+			emit(key, value)
+			i++
+		}
+	}, beam.Impulse(s), beam.SideInput{Input: src})
+
+	beam.ParDo(s, &load.RuntimeMonitor{}, src)
+
+	if err := beamx.Run(ctx, p); err != nil {
```
Comment (kamilwu) [inline sdks/go/test/load/sideinput/sideinput.go:79]: `beamx.Run` -> `beamx.RunWithMetrics`. Please also add the code for sending results to the database.

File: .test-infra/jenkins/job_LoadTests_SideInput_Flink_Go.groovy (hunk 1 above)
Comment (tszerszen) [reply .test-infra/jenkins/job_LoadTests_SideInput_Flink_Go.groovy]: Thank you, in the last commit I removed  ${now} from influx_measurement.

File: sdks/go/test/load/sideinput/sideinput.go (hunk 3 above)
Comment (tszerszen) [reply sdks/go/test/load/sideinput/sideinput.go]: Ok, now it's run with beamx.RunWithMetrics.

File: sdks/go/test/load/sideinput/sideinput.go (hunk 2 above)
Comment (tszerszen) [reply sdks/go/test/load/sideinput/sideinput.go]: I removed default value for flag syntheticSourceConfig.

**Comment by tszerszen** [comment]
Run Spotless PreCommit

File: .test-infra/jenkins/job_LoadTests_SideInput_Flink_Go.groovy [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        influx_namespace   : 'flink',
+        influx_measurement : 'go_batch_sideinput_1',
+        input_options      : '\'{' +
+        '"num_records": 10000000,' +
+        '"key_size": 100,' +
+        '"value_size": 900}\'',
+        parallelism        : 10,
+        endpoint           : 'localhost:8099',
+        environment_type   : 'DOCKER',
+        environment_config : "${DOCKER_CONTAINER_REGISTRY}/beam_go_sdk:latest",
```
Comment (kamilwu) [inline .test-infra/jenkins/job_LoadTests_SideInput_Flink_Go.groovy:44]: This test is exactly the same as one below. Did you miss `access_elements`?

File: .test-infra/jenkins/job_LoadTests_SideInput_Flink_Go.groovy [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+def batchScenarios = {
+  [
+    [
+      title          : 'SideInput Go Load test: 10gb-1kb-10workers-1window-first-iterable',
+      test           : 'sideinput',
+      runner         : CommonTestProperties.Runner.FLINK,
+      pipelineOptions: [
+        job_name           : "load-tests-go-flink-batch-sideinput-1-${now}",
+        influx_namespace   : 'flink',
+        influx_measurement : 'go_batch_sideinput_1',
```
Comment (kamilwu) [inline .test-infra/jenkins/job_LoadTests_SideInput_Flink_Go.groovy:36]: You _cannot_ use `1` here. This number is an ID that uniquely identifies the test. For side input tests:
* `1` corresponds to the `1gb-1kb-10workers-1window-1key-percent-dict` test
* `2` corresponds to the `1gb-1kb-10workers-1window-99key-percent-dict` test
and so on and so forth. You can see a full set of side input tests here: https://github.com/apache/beam/blob/master/.test-infra/jenkins/job_LoadTests_SideInput_Python.groovy

File: .test-infra/jenkins/job_LoadTests_SideInput_Flink_Go.groovy (hunk 4 above)
Comment (tszerszen) [reply .test-infra/jenkins/job_LoadTests_SideInput_Flink_Go.groovy:44]: @kamilwu if you look at `job_LoadTests_SideInput_Python.groovy` there are two iterable tests. They are exactly the same with the exception of a name, here's the fragment of code:
```groovy
    [
      name: '10gb-1kb-10workers-1window-first-iterable',
      testSpecificOptions: [
        input_options    : '\'{' +
        '"num_records": 10000000,' +
        '"key_size": 100,' +
        '"value_size": 900}\'',
        side_input_type  : 'iter',
        access_percentage: 1,
      ]
    ],
    [
      name: '10gb-1kb-10workers-1window-iterable',
      testSpecificOptions: [
        input_options    : '\'{' +
        '"num_records": 10000000,' +
        '"key_size": 100,' +
        '"value_size": 900}\'',
        side_input_type  : 'iter',
      ]
    ],
```
And default access_percentage is 1, and it's not defined in second test case.

File: .test-infra/jenkins/job_LoadTests_SideInput_Flink_Go.groovy (hunk 5 above)
Comment (tszerszen) [reply .test-infra/jenkins/job_LoadTests_SideInput_Flink_Go.groovy]: Thank you, you're right. Just fixed it.

File: .test-infra/jenkins/job_LoadTests_SideInput_Flink_Go.groovy (hunk 4 above)
Comment (tszerszen) [reply .test-infra/jenkins/job_LoadTests_SideInput_Flink_Go.groovy:45]: @kamilwu  Sorry, you're right default access_percentage is 100, therefore first test should have access_percentage 1, and second one none, which means it gets the default 100.

File: sdks/go/test/load/sideinput/sideinput.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+	if *syntheticSourceConfig == "" {
+		panic("--input_options not provided")
+	} else {
+		encoded := []byte(*syntheticSourceConfig)
+		return synthetic.DefaultSourceConfig().BuildFromJSON(encoded)
+	}
+}
+
+type doFn struct {
+	elementsToAccess int
```
Comment (lostluck) [inline sdks/go/test/load/sideinput/sideinput.go:55]: This isn't the main problem, we're debugging, but DoFn fields *must* be exported to get serialized properly.

eg. ElementsToAccess instead.

**Comment by lostluck** [comment]
WRT the investigation we've been doing around a flink failure for large settings: 
In my own investigation into the problem on the google internal runner, the "big" configuration ends up being too large for protocol buffer serialization, with a single ~10GB StateResponse, causing the issue. Protos have a hard cap of 2GB in serialized size.

From #beam-go slack discussion, python also fails with large configurations..

`--input_options='{"num_records": 2000000, "key_size":100, "value_size":900}' --access_percentage=1`
and
`--input_options='{"num_records": 10000000, "key_size":100, "value_size":180}' --access_percentage=1`
work by cutting data total down to 1/5th.

This will affect all portable runners, but not any of the legacy ones, because they don't pass data around through the protos. A JIRA is being filed about it.

File: sdks/go/test/load/sideinput/sideinput.go (hunk 6 above)
Comment (tszerszen) [reply sdks/go/test/load/sideinput/sideinput.go]: Thank you, I changed the name to start with capital letter.

**Comment by tszerszen** [comment]
R: @kamilwu ?

File: .test-infra/jenkins/job_LoadTests_SideInput_Go.groovy [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+import LoadTestsBuilder as loadTestsBuilder
+import PhraseTriggeringPostCommitBuilder
+import InfluxDBCredentialsHelper
+
+String now = new Date().format('MMddHHmmss', TimeZone.getTimeZone('UTC'))
+
+def batchScenarios = {
+  [
+    [
+      title          : 'SideInput Go Load test: 400mb-1kb-10workers-1window-first-iterable',
```
Comment (kamilwu) [inline .test-infra/jenkins/job_LoadTests_SideInput_Go.groovy:30]: In the Flink job, the title begins with `10gb`, while here it begins with `400mb`. Please choose one of these and stick to it.

File: .test-infra/metrics/grafana/dashboards/perftests_metrics/SideInput_Load_Tests.json [hunk 8]
```diff
@@ -21,7 +21,7 @@
   "links": [],
   "panels": [
     {
-      "content": "The following options are common to all tests:\n* key size: 100B\n* value size: 900B\n* number of workers: 10\n* size of the window (if fixed windows are used): 1 second\n\nAdditional common options for Dataflow:\n* experiments: use_runner_v2\n* autoscaling_algorithm: NONE\n\n\n[Jenkins job definition (Python, Dataflow)](https://github.com/apache/beam/blob/master/.test-infra/jenkins/job_LoadTests_SideInput_Python.groovy)\n\n",
+      "content": "The following options are common to all tests:\n* key size: 100B\n* value size: 900B\n* number of workers: 10\n* size of the window (if fixed windows are used): 1 second\n\nAdditional common options for Dataflow:\n* experiments: use_runner_v2\n* autoscaling_algorithm: NONE\n\n\n[Jenkins job definition (Python, Dataflow)](https://github.com/apache/beam/blob/master/.test-infra/jenkins/job_LoadTests_SideInput_Python.groovy)\n\nUntil the issue [BEAM-11427](https://issues.apache.org/jira/browse/BEAM-11427) in Go SDK is resolved, sideinput iteration test have 400MB, instead of 10GB.",
```
Comment (kamilwu) [inline .test-infra/metrics/grafana/dashboards/perftests_metrics/SideInput_Load_Tests.json:24]: The comment is OK, but you should also specify additional options for Flink. And while there is a link to Python Jenkins job, you could also add links to your newly defined jobs.

**Review by kamilwu — COMMENTED** [review]
Thanks, we're almost done.

File: .test-infra/jenkins/job_LoadTests_SideInput_Go.groovy (hunk 7 above)
Comment (tszerszen) [reply .test-infra/jenkins/job_LoadTests_SideInput_Go.groovy:30]: Thank I changed everything to 400mb-*.

**Review by kamilwu — APPROVED** [review]

**Pull request merged**